mineral-ui/mineral-ui
# fix(textinput): Fix #609 lose focus after typing

### Description

Fixes a regression in TextInput wherein the input would immediately lose focus after typing into it.

### Motivation and context

Fixes #609 

The problem was that the control node, an input in this case, was being defined in `render()` and the DOM element thus was being recreated and completely replaced whenever anything changed, and thus losing focus.  The solution was to create the node outside of render and only update it when the control prop changes.  We follow this pattern elsewhere in the codebase.

For future reference, the pattern to avoid here is calling `createStyledComponent` inside `render()` whenever focus on that element is potentially important.

FWIW, I tried to additionally write a regression test in TextInput for this issue but was unable to create a failing test as triggering change and/or keydown events in jest/enzyme did not cause focus to be lost.  I could create one on FauxControl to validate that `createControlNode` is only called when `control` changes, but that seemed rather limited in value.  Open to suggestions...

### Screenshots, videos, or demo, if appropriate

http://609-textinput-focus--mineral-ui.netlify.com/components/text-input/controlled/

### How to test

1. View the the demo in the link above
2. Type into the TextInput and ensure that focus is not lost.

_Note that due to an issue in our example, tabbing out of the input causes the whole example to be re-rendered - re-invokes the constructor, thus resetting the value back to the default.  It is unrelated to this issue._ 

### Types of changes

- Bug fix (non-breaking change which fixes an issue)

### Checklist

* [x] Automated tests written and passing - Existing coverage.  Tried to write regression test, but was unsuccessful.  Open to suggestions.
